File: Leetcode Challenge/February/cpp/Maximum Width of Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int widthOfBinaryTree(TreeNode* root) {
        queue<pair<TreeNode*, int>> q;
        q.push({root, 0});
        long ans = 0;
        while(!q.empty()){
            int n = q.size();
            long start = q.front().second;
            long end = q.back().second;
            ans = max(ans, end-start+1);
            for(int i = 0; i < n; i++) {
                pair<TreeNode*, int> p = q.front(); 
                TreeNode* node = p.first;
                long idx = p.second;
                q.pop();
                if(node->left) q.push({node->left, 2*idx+1});
                if(node->right) q.push({node->right, 2*idx+2});
            }
        }
        return ans;
    }
};

File: Leetcode Challenge/February/Python/Maximum Width of Binary Tree.py
class Solution(object):
    def widthOfBinaryTree(self, root):
        
        if root==None:
            return 0
        
        ans=1
        q=[[root,0]]
        
        while len(q)>0:
            size=len(q)
            start=q[0][1]
            end=q[-1][1]
            ans=max(ans,end-start+1)
            for i in range(size):
                current=q[0]
                idx=current[1]-start
                q.pop(0)
                if current[0].left!=None:
                    q.append([current[0].left,2*idx+1])
                if current[0].right!=None:
                    q.append([current[0].right,2*idx+2])
                    
            
        return ans  
